File: Scadex.Model/Dtos/Realtime/Queries/RealtimeEvents.cs
using Scadex.Core.Model;
using Scadex.Model.Enums;

namespace Scadex.Model.Dtos.Realtime.Queries;

public class ChannelValueChange : IDto
{
    public Guid IoChannelId { get; set; }
    public Guid DeviceId { get; set; }
    public int ChannelNumber { get; set; }
    public string? Value { get; set; }
    public DateTime UpdatedAt { get; set; }
}

public class DeviceStatusChange : IDto
{
    public Guid DeviceId { get; set; }
    /// <summary>Null = hic telemetri alinmadi. <c>Offline</c> ile AYNI SEY DEGIL.</summary>
    public EntityEnums.DeviceStatus? StatusId { get; set; }
    public DateTime? LastSeen { get; set; }
}

public class CabinetStatusChange : IDto
{
    public Guid CabinetId { get; set; }
    public EntityEnums.DeviceStatus? StatusId { get; set; }
    public DateTime? LastSeen { get; set; }
    public DateTime? ScadaLastIngestAt { get; set; }
}

/// <summary>
/// Komutu gonderen zaten HTTP yanitinda sonucu aliyor ancak bu event ayni kabini izleyen DIGER kullanicilar da giden komutun cevabını görebilsindel diye var.
/// Kanal DEGERI bu olayla gelmez: SCADA komutu uyguladiginda degisen deger normal ingest yoluyla <see cref="ChannelValueChange"/> olarak gelir.
/// </summary>
public class CommandCompleted : IDto
{
    public Guid CommandId { get; set; }
    public Guid DeviceId { get; set; }
    public Guid? IoChannelId { get; set; }
    public int? ChannelNumber { get; set; }
    public EntityEnums.DeviceCommandType CommandType { get; set; }
    public EntityEnums.CommandStatus Status { get; set; }
    public string? ResultMessage { get; set; }
    public DateTime? RespondedAt { get; set; }
    public string? RequestedByName { get; set; }
}

File: Scadex.WebUI/src/models/realtime/queries/realtimeEvents.ts
/** Ayna: Scadex.Model/Dtos/Realtime/Queries/RealtimeEvents.cs — sözleşme: docs/api-contract/09-realtime.md */
import type { CommandStatus, DeviceCommandType, DeviceStatus } from '@/models/enums';

/**
 * `/hubs/diagram` üzerinden sunucudan gelen olayların gövdeleri.
 *
 * Hub olayları REST yanıtlarıyla **aynı** serializer'dan geçiyor
 * (`AddJsonProtocol` + `ApiJsonOptions.Apply`), dolayısıyla casing ve sayısal enum
 * kodlaması buradaki diğer DTO aynalarıyla birebir aynı.
 */

export interface ChannelValueChange {
  ioChannelId: string;
  /** Kanalın bağlı olduğu cihaz — node bazında yeniden çizim için. */
  deviceId: string;
  channelNumber: number;
  /** Null = kanal var ama okunamadı. `"0"` ile AYNI ŞEY DEĞİL. */
  value: string | null;
  updatedAt: string;
}

export interface DeviceStatusChange {
  deviceId: string;
  /** Null = hiç telemetri alınmadı. `Offline` (0) ile AYNI ŞEY DEĞİL. */
  statusId: DeviceStatus | null;
  lastSeen: string | null;
}

export interface CabinetStatusChange {
  cabinetId: string;
  statusId: DeviceStatus | null;
  lastSeen: string | null;
  scadaLastIngestAt: string | null;
}

/**
 * Bir kumandanın sonuçlandığı bildirimi.
 *
 * Üstteki üç olaydan **farklı bir sebeple** var: onlar sahadan gelen değişimi
 * taşır, bu ise bir kullanıcının yaptığı işi. Komutu gönderen sonucu zaten HTTP
 * yanıtında alır; bu yayın aynı kabini izleyen **diğerleri** için.
 *
 * **Gönderen de bu olayı alır.** Sunucu grubun tamamına yayınlıyor ve gönderenin
 * bağlantısı da o grupta; ayıklama istemcide, `commandId` üzerinden yapılır.
 *
 * Gövde `DeviceCommandResultDto`'nun ALT KÜMESİDİR (`payloadJson`, `sentAt`,
 * `elapsedMs` yok) — bu yüzden doğrudan geçmiş listesine eklenmez.
 */
export interface CommandCompleted {
  commandId: string;
  deviceId: string;
  ioChannelId: string | null;
  channelNumber: number | null;
  commandType: DeviceCommandType;
  status: CommandStatus;
  resultMessage: string | null;
  respondedAt: string | null;
  requestedByName: string | null;
}

/**
 * Sunucunun çağırdığı metot ADLARI. Sözleşmenin en kırılgan parçası: yanlış
 * yazılan bir ad hiçbir hata üretmez, olay sessizce hiçbir yere gitmez ve sorun
 * ancak arayüzde "canlı veri gelmiyor" olarak fark edilir.
 *
 * Tek yerde sabitlenmeleri, en azından yazım hatasının tek bir yerde olmasını
 * sağlar ve dinleyici tarafında derleyici desteği verir.
 */
export const DiagramHubEvents = {
  channelValuesChanged: 'ChannelValuesChanged',
  deviceStatusChanged: 'DeviceStatusChanged',
  cabinetStatusChanged: 'CabinetStatusChanged',
  commandCompleted: 'CommandCompleted'
} as const;

/** İstemcinin çağırdığı hub metotları. */
export const DiagramHubMethods = {
  subscribe: 'Subscribe',
  unsubscribe: 'Unsubscribe'
} as const;
